extension/observer/hostobserver/extension.go

```diff
@@ -0,0 +1,249 @@
+				logger: logger,
+			},
+		},
+	}
+
+	return h, nil
+}
+
+func (h *hostObserver) Start(context.Context, component.Host) error {
+	return nil
+}
+
+func (h *hostObserver) Shutdown(context.Context) error {
+	h.StopListAndWatch()
+	return nil
+}
+
+func (e endpointsLister) ListEndpoints() []observer.Endpoint {
+	conns, err := getConnections()
+	if err != nil {
+		e.logger.Error("Could not get local network listeners", zap.Error(err))
+		return nil
+	}
+
+	return e.collectEndpoints(conns)
+}
+
+var getConnections = func() (conns []net.ConnectionStat, err error) {
+	// Skip UID lookup since it's not used by the observer, the method
+	// is available only on linux. See https://github.com/shirou/gopsutil/pull/783
+	// for details.
+	if runtime.GOOS == "linux" {
+		conns, err = net.ConnectionsWithoutUids("all")
+	} else {
+		conns, err = net.Connections("all")
+	}
+
+	return conns, err
+}
+
+func (e endpointsLister) collectEndpoints(conns []net.ConnectionStat) []observer.Endpoint {
+	endpoints := make([]observer.Endpoint, 0, len(conns))
+	connsByPID := make(map[int32][]*net.ConnectionStat)
+	for i := range conns {
+		c := conns[i]
+		isIPSocket := c.Family == syscall.AF_INET || c.Family == syscall.AF_INET6
+		isTCPOrUDP := c.Type == syscall.SOCK_STREAM || c.Type == syscall.SOCK_DGRAM
+		// UDP doesn't have any status
+		isUDPOrListening := c.Type == syscall.SOCK_DGRAM || c.Status == "LISTEN"
+		// UDP is "listening" when it has a remote port of 0
+		isTCPOrHasNoRemotePort := c.Type == syscall.SOCK_STREAM || c.Raddr.Port == 0
+
+		if !isIPSocket || !isTCPOrUDP || !isUDPOrListening || !isTCPOrHasNoRemotePort {
+			continue
+		}
+
+		// PID of 0 means that the listening file descriptor couldn't be mapped
+		// back to a process's set of open file descriptors in /proc. Collect these
+		// endpoints even though there's no process metadata available so users can
+		// still do discovery rules on such sockets.
+		if c.Pid == 0 {
+			cd := collectConnectionDetails(c)
+			id := observer.EndpointID(fmt.Sprintf("%s-%d-%s", c.Laddr.IP, cd.port, cd.transport))
+
+			endpoints = append(endpoints, observer.Endpoint{
+				ID:     id,
+				Target: cd.target,
+				Details: observer.HostPort{
+					Name:      string(id),
+					Port:      cd.port,
+					Transport: cd.transport,
+					// TODO: Move this field to observer.Endpoint and
+					// update receiver_creator to filter IPv4/IPv6.
+					IsIPv6: cd.isIPv6,
+				},
+			})
+			continue
+		}
+
+		connsByPID[c.Pid] = append(connsByPID[c.Pid], &c)
+	}
+
+	for pid, conns := range connsByPID {
+		proc, err := process.NewProcess(pid)
+		if err != nil {
+			e.logger.Warn("Could not examine process (it might have terminated already)")
+			continue
+		}
+
+		pd, err := collectProcessDetails(proc)
+		if err != nil {
+			e.logger.Error(
+				"Skipping process",
+				zap.Int32("pid", pid),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		for _, c := range conns {
+			cd := collectConnectionDetails(*c)
```

Comment:
If it's already a pointer probably better to just pass as a pointer instead of making a copy.